presto-memory-context/src/main/java/com/facebook/presto/memory/AggregatedMemoryContext.java

```diff
@@ -0,0 +1,161 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.memory;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.util.concurrent.ListenableFuture;
+import io.airlift.units.DataSize;
+
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.GuardedBy;
+import javax.annotation.concurrent.ThreadSafe;
+
+import static com.facebook.presto.memory.MemoryReservationHandler.NOT_BLOCKED;
+import static com.google.common.base.MoreObjects.toStringHelper;
+import static io.airlift.units.DataSize.Unit.MEGABYTE;
+
+@ThreadSafe
+public class AggregatedMemoryContext
+{
+    public static final long GUARANTEED_MEMORY = new DataSize(1, MEGABYTE).toBytes();
+
+    @Nullable
+    protected final AggregatedMemoryContext parentMemoryContext;
+    @Nullable
+    private final MemoryReservationHandler reservationHandler;
+    @GuardedBy("this")
+    protected long usedBytes;
+    @GuardedBy("this")
+    private boolean closed;
+
+    public AggregatedMemoryContext()
+    {
+        this(null, null);
+    }
+
+    public AggregatedMemoryContext(MemoryReservationHandler reservationHandler)
+    {
+        this(null, reservationHandler);
+    }
+
+    private AggregatedMemoryContext(AggregatedMemoryContext parentMemoryContext, MemoryReservationHandler reservationHandler)
+    {
+        this.parentMemoryContext = parentMemoryContext;
```

Comment:
assert that at least one of `parentMemoryContext` and `reservationHandler` is null.